tests/webhooksreceiver.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2023 Sorint.lab
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tests
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sync"
+	"testing"
+
+	"agola.io/agola/internal/services/notification"
+	"agola.io/agola/internal/testutil"
+	"agola.io/agola/internal/util"
+	"github.com/gorilla/mux"
+	"github.com/mitchellh/copystructure"
+	"github.com/rs/zerolog"
+)
+
+type webhooksReceiver struct {
+	ctx     context.Context
+	cancel  context.CancelFunc
+	started bool
+
+	log zerolog.Logger
+
+	t   *testing.T
+	dir string
+
+	mu sync.Mutex
+
+	listenAddress string
+	exposedURL    string
+
+	webhooksStatus *webhooksStatus
+}
+
+type webhooksStatus struct {
+	webhooks []*notification.Webhook
+	mu       sync.Mutex
+}
+
+func (ws *webhooksStatus) getWebhooks() []*notification.Webhook {
+	ws.mu.Lock()
+	defer ws.mu.Unlock()
+
+	return ws.webhooks
```

Comment:
The lock is unuseful if you return the slice. You should return a deecopy of ws.webhooks.